lib/pacemaker/pcmk_sched_allocate.c

```diff
@@ -2393,6 +2392,121 @@ order_first_probes(pe_working_set_t * data_set)
     }
 }
 
+static void
+order_first_probe_then_restart_repromote(pe_action_t * probe, pe_action_t * after)
+{
+    GListPtr gIter = NULL;
+
+    if (probe == NULL || safe_str_neq(probe->task, RSC_STATUS)) {
+        return;
+    }
+
+    if (after == NULL) {
+        return;
+    }
+
+    if (probe->rsc == NULL) {
+        return;
+    }
+
+    /* Avoid running into any possible loop */
+    if (is_set(after->flags, pe_action_tracking)) {
+        return;
+    }
+
+    pe_set_action_bit(after, pe_action_tracking);
+
+    crm_trace("Processing based on %s %s -> %s %s",
+              probe->uuid,
+              probe->node ? probe->node->details->uname: "",
+              after->uuid,
+              after->node ? after->node->details->uname : "");
+
+    if (after->rsc
+        && probe->rsc != after->rsc
+        && uber_parent(probe->rsc) != after->rsc) {
+
+        GListPtr then_actions = NULL;
+        enum pe_ordering probe_order_type = pe_order_optional;
+
+        if (safe_str_eq(after->task, RSC_START)) {
+            then_actions = pe__resource_actions(after->rsc, NULL, RSC_STOP, FALSE);
+
+        } else if (safe_str_eq(after->task, RSC_PROMOTE)) {
+            then_actions = pe__resource_actions(after->rsc, NULL, RSC_DEMOTE, FALSE);
+        }
+
+        for (gIter = then_actions; gIter != NULL; gIter = gIter->next) {
+            pe_action_t *then = (pe_action_t *) gIter->data;
+
+            order_actions(probe, then, probe_order_type);
+        }
+        g_list_free(then_actions);
+    }
+
+    for (gIter = after->actions_after; gIter != NULL; gIter = gIter->next) {
+        pe_action_wrapper_t *after_wrapper = (pe_action_wrapper_t *) gIter->data;
+
+        if (is_set(after_wrapper->type, pe_order_implies_then)) {
+            order_first_probe_then_restart_repromote(probe, after_wrapper->action);
+        }
```

Comment:
Why limit this to pe_order_implies_then?

Reply:
pe_order_implies_then is the reason why a required A.start implies/enforces B.start to be required too, which is the cause of B.restart/re-promote. Without limiting to pe_order_implies_then, it'd trace/introduce unnecessary ordering or even transition loops... 

I'm not sure about pe_order_implies_then_on_node though. It's now only used for unfencing case, which tends to introduce transition loops....

Do you think of other suitable cases?